### PR TITLE
optimize memory allocation

### PR DESCRIPTION
**Pull request by Theodoree**
# PR Details

1.align xlsxRow struct.
2.optimization workSheetWriter
3. add NewSheetWithRowNum 

## Description

Mainly to optimize memory allocation

## Related Issue
l wish somebody can optimize xlsxC struct. it's too bigger.Maybe it can be optimized smaller.

## Motivation and Context

sometime l use this package. memory will alloc 500m ~ 1000m. it's too expensive

## How Has This Been Tested

```

go test -run=none -bench=BenchmarkNewSheetWithRowNum -benchmem -count 10

name                   old time/op    new time/op    delta
NewSheetWithRowNum1-8     667µs ± 5%     679µs ±15%     ~     (p=0.968 n=9+10)

name                   old alloc/op   new alloc/op   delta
NewSheetWithRowNum1-8     643kB ± 0%     588kB ± 0%   -8.46%  (p=0.000 n=10+10)

name                   old allocs/op  new allocs/op  delta
NewSheetWithRowNum1-8     10.3k ± 0%      9.2k ± 0%  -10.68%  (p=0.000 n=10+10)


go test -run=none -bench=BenchmarkFile_SaveAs -benchmem -count 10
name           old time/op    new time/op    delta
File_SaveAs-8     738µs ±13%     698µs ±10%     ~     (p=0.105 n=10+10)

name           old alloc/op   new alloc/op   delta
File_SaveAs-8     643kB ± 0%     588kB ± 0%   -8.46%  (p=0.000 n=10+10)

name           old allocs/op  new allocs/op  delta
File_SaveAs-8     10.3k ± 0%      9.2k ± 0%  -10.67%  (p=0.000 n=9+10)



```

## Types of changes

- [ x ] New feature (non-breaking change which adds functionality)
- [ x ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
- [ x ] I have added tests to cover my changes.
- [ x ] All new and existing tests passed.

### CODE REVIEW

File: xmlWorksheet.go [hunk 1]
```diff
@@ -313,20 +313,20 @@ type xlsxSheetData struct {
 // xlsxRow directly maps the row element. The element expresses information
 // about an entire row of a worksheet, and contains all cell definitions for a
 // particular row in the worksheet.
-type xlsxRow struct {
+type xlsxRow struct {	// alignment word
```
Comment (xuri) [inline xmlWorksheet.go:316]: struct fields order should be consistent with defined in XSD with ECMA-376 Spec: [`CT_Row`](https://github.com/xuri/xsd/blob/b7fd4e5c7e0d2ad785de32e140161ea5782f5d24/xsd/OfficeOpenXML-XMLSchema-Strict/sml.xsd#L2238-L2255).

**Review by xuri — CHANGES_REQUESTED** [review]

File: sheet.go [hunk 2]
```diff
@@ -64,6 +64,19 @@ func (f *File) NewSheet(name string) int {
 	return f.GetSheetIndex(name)
 }
 
+// NewSheetWithRowNum provides the function to create a new sheet by given a worksheet and a given number of rows.
```
Comment (xuri) [inline sheet.go:67]: This new API seems unnecessary, SetCellValue will preallocate cells for any new worksheets.

Comment (Theodoree) [reply sheet.go:67]: Sometime will need this. Realloc cause huge GC pressure.
Example:
I need export a million row form db and write into excel, data is encrypt , need decrypt . 
slice often realloc we cause huge GC pressure 、copy spending

File: xmlWorksheet.go (hunk 1 above)
Comment (Theodoree) [reply xmlWorksheet.go:316]: done

File: sheet.go (hunk 2 above)
Comment (Theodoree) [reply sheet.go:67]: I have encountered this situation before.
l remember alloc 1-3G mem. it's too Impressive.

**Comment by Theodoree** [comment]
@xuri

File: sheet.go (hunk 2 above)
Comment (xuri) [reply sheet.go:67]: Sorry for my late reply. We can using streaming API [`NewStreamWriter`](https://pkg.go.dev/github.com/360EntSecGroup-Skylar/excelize/v2#File.NewStreamWriter) to generate huge data spreadsheet, by evaluation and follow the principle of least availability, I don't purpose to add this API. If you have any questions, let me know.

**Comment by Theodoree** [comment]
go test -run=none -bench=BenchmarkNewSheet -benchmem -count 10

name                        time/op
NewSheetWithRowNum-8         631µs ±12%
NewSheetWithOutRowNum-8      636µs ± 2%
NewSheetWithStreamWriter-8  1.07ms ± 3%

name                        alloc/op
NewSheetWithRowNum-8         508kB ± 0%
NewSheetWithOutRowNum-8      616kB ± 0%
NewSheetWithStreamWriter-8   880kB ± 4%

name                        allocs/op
NewSheetWithRowNum-8         9.18k ± 0%
NewSheetWithOutRowNum-8      9.20k ± 0%
NewSheetWithStreamWriter-8   12.6k ± 0%

**Comment by Theodoree** [comment]
> go test -run=none -bench=BenchmarkNewSheet -benchmem -count 10
> 
> name time/op
> NewSheetWithRowNum-8 631µs ±12%
> NewSheetWithOutRowNum-8 636µs ± 2%
> NewSheetWithStreamWriter-8 1.07ms ± 3%
> 
> name alloc/op
> NewSheetWithRowNum-8 508kB ± 0%
> NewSheetWithOutRowNum-8 616kB ± 0%
> NewSheetWithStreamWriter-8 880kB ± 4%
> 
> name allocs/op
> NewSheetWithRowNum-8 9.18k ± 0%
> NewSheetWithOutRowNum-8 9.20k ± 0%
> NewSheetWithStreamWriter-8 12.6k ± 0%

**Comment by Theodoree** [comment]
l will Think about it, maybe I'll delete this api.  `NewSheetWithRowNum`

File: sheet_test.go [hunk 3]
```diff
@@ -344,3 +345,91 @@ func TestSetSheetName(t *testing.T) {
...
+}
+func newSheetWithRowNum() {
+	file := NewFile()
+
+	file.NewSheetWithRowNum("sheet1", 1000)
+
+	for i := 0; i < 1000; i++ {
+		file.SetCellInt("sheet1", "A"+strconv.Itoa(i+1), i)
+	}
+	file = nil
```
Comment (xuri) [inline sheet_test.go:390]: The sheet data doesn't save in `newSheetWithRowNum` and `newSheetWithOutRowNum`, but `newSheetWithStreamWriter`, so these profiling benchmarks are unfair comparison.

**Comment by xuri** [comment]
FYI, [here is](https://xuri.me/excelize/zh-hans/performance.html) a graph that shows the performance comparison of normal API and streaming writing API. I'll public testing code later.

File: sheet_test.go (hunk 3 above)
Comment (Theodoree) [reply sheet_test.go:390]: sry.I didn't notice.

**Comment by Theodoree** [comment]
@xuri  L have deleted that api. l was testing. l find `NewStreamWriter `  will save more mem.but Execution time will be longer.

**Comment by xuri** [comment]
Usually, the streaming API will use less memory and faster than normal API, but affected by business scenario, you need trade-off by yourself. Try profiling your program by `pprof` to get more details.

**Comment by Theodoree** [comment]
@xuri  so.when can merge PR ?

**Pull request merged**